src/Storages/MergeTree/MergeTreeIndexHnsw.cpp

```diff
@@ -0,0 +1,351 @@
+{
+    /// Number of dimensions is required in the index constructor,
+    /// so it must be written and read separately from the other part
+    writeIntBinary(static_cast<UInt64>(index->getDimensions()), ostr); // write dimension
+    index->serialize(ostr);
+}
+
+template <unum::usearch::metric_kind_t Metric>
+void MergeTreeIndexGranuleUSearch<Metric>::deserializeBinary(ReadBuffer & istr, MergeTreeIndexVersion /*version*/)
+{
+    UInt64 dimension;
+    readIntBinary(dimension, istr);
+    index = std::make_shared<USearchIndexWithSerialization<Metric>>(dimension);
+    index->deserialize(istr);
+}
+
+template <unum::usearch::metric_kind_t Metric>
+MergeTreeIndexAggregatorUSearch<Metric>::MergeTreeIndexAggregatorUSearch(const String & index_name_, const Block & index_sample_block_)
+    : index_name(index_name_), index_sample_block(index_sample_block_)
+{
+}
+
+template <unum::usearch::metric_kind_t Metric>
+MergeTreeIndexGranulePtr MergeTreeIndexAggregatorUSearch<Metric>::getGranuleAndReset()
+{
+    auto granule = std::make_shared<MergeTreeIndexGranuleUSearch<Metric>>(index_name, index_sample_block, index);
+    index = nullptr;
+    return granule;
+}
+
+template <unum::usearch::metric_kind_t Metric>
+void MergeTreeIndexAggregatorUSearch<Metric>::update(const Block & block, size_t * pos, size_t limit)
+{
+    if (*pos >= block.rows())
+        throw Exception(
+            ErrorCodes::LOGICAL_ERROR,
+            "The provided position is not less than the number of block rows. Position: {}, Block rows: {}.",
+            *pos,
+            block.rows());
+
+    size_t rows_read = std::min(limit, block.rows() - *pos);
+    if (rows_read == 0)
+        return;
+
+    if (index_sample_block.columns() > 1)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Expected block with single column");
+
+    const String & index_column_name = index_sample_block.getByPosition(0).name;
+    ColumnPtr column_cut = block.getByName(index_column_name).column->cut(*pos, rows_read);
+
+    if (const auto & column_array = typeid_cast<const ColumnArray *>(column_cut.get()))
+    {
+        const auto & data = column_array->getData();
+        const auto & array = typeid_cast<const ColumnFloat32 &>(data).getData();
+
+        if (array.empty())
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Array has 0 rows, {} rows expected", rows_read);
+
+        const auto & offsets = column_array->getOffsets();
+        const size_t num_rows = offsets.size();
+
+
+        /// Check all sizes are the same
+        size_t size = offsets[0];
+        for (size_t i = 0; i < num_rows - 1; ++i)
+            if (offsets[i + 1] - offsets[i] != size)
+                throw Exception(ErrorCodes::INCORRECT_DATA, "All arrays in column {} must have equal length", index_column_name);
+
+
+        index = std::make_shared<USearchIndexWithSerialization<Metric>>(size);
+
+        /// Add all rows of block
+        index->reserve(unum::usearch::ceil2(index->size() + num_rows + 1));
```

Comment:
Is `reseve()` mandatory? (in that case, `add` could do a corresponding check as a precondition).

Reply:
In the current implementation, `add` won't pass unless the capacity is reserved beforehand. This is done because USearch is a concurrent thread-safe structure, which should remain adequate even if the insertion fails halfway. If we introduce memory allocations there, the procedure becomes way too convoluted.

Reply:
It's guaranteed in ClickHouse at least that errors (exception) during insertion into the index are propagated to the user, abort index creation and generally destroy all (half-baked) index structures. So at least in ClickHouse, there is no need for thread-safe insert.

But I understand that reserving memory beforehand simplifies things, we can keep it as is.